webapp/src/js/services/auth.js

```diff
@@ -25,71 +25,75 @@ angular.module('inboxServices').factory('Auth',
     'ngInject';
     'use strict';
 
-    var check = function(permissions, userRoles, settings, expected) {
-      return _.every(permissions, function(permission) {
-        var roles = settings.permissions[permission];
+    let failureReason;
+
+    const check = (permissions, userRoles, settings, expected) => {
+      return _.every(permissions, (permission) => {
+        const roles = settings.permissions[permission];
         if (!roles) {
           return !expected;
         }
-        var found = _.intersection(userRoles, roles).length > 0;
+        const found = _.intersection(userRoles, roles).length > 0;
         return expected === found;
       });
     };
 
-    var isRequired = function(permission) {
+    const isRequired = (permission) => {
       return permission.indexOf('!') !== 0;
     };
 
-    var getRequired = function(permissions) {
+    const getRequired = (permissions) => {
       return _.filter(permissions, isRequired);
     };
 
-    var getDisallowed = function(permissions) {
+    const getDisallowed = (permissions) => {
       permissions = _.reject(permissions, isRequired);
-      permissions = _.map(permissions, function(permission) {
+      permissions = _.map(permissions, (permission) => {
         return permission.substring(1);
       });
       return permissions;
     };
 
-    var authFail = function(reason, permissions, roles) {
-      $log.debug('Auth failed: ' + reason + '. User roles: ' + roles + '. Wanted permissions: ' + permissions);
+    const authFail = (reason, permissions, roles) => {
+      $log.debug(`Auth failed: ${reason}. User roles: ${roles}. Wanted permissions: ${permissions}`);
       return $q.reject();
     };
 
-    var checkPermissions = function(required, disallowed, roles, settings) {
+    const checkPermissions = (required, disallowed, roles, settings) => {
       if (!check(required, roles, settings, true)) {
-        return 'missing required permission(s)';
+        failureReason = 'missing required permission(s)';
```

Comment:
Don't store state unless you have to. This is dangerous because the previous failure reason isn't being cleared when a new request comes in. Also it hangs around in memory forever which is unnecessary.

The way this function worked previously looks good to me! Why did you change it?

Some people are uncomfortable with how JS can return different types (in this case, a string or a boolean). If that's the case you could try returning an object, eg:

```js
return { success: false, message: 'missing required permission(s)' };
```

This is a bit more verbose than the original but always returns an object with the same structure.

The important thing is not to store state outside of this scope when it's not relevant to any other scope.